File: go/indicators/johnehlers/hilberttransformer/phaseaccumulatorestimator.go
package hilberttransformer

import (
	"math"
	"sync"
)

const accumulationLength = 40

// PhaseAccumulatorEstimatorEstimator implements the Hilbert transformer of the
// WMA-smoothed and detrended data followed by the Phase Accumulation to determine
// the instant period.
//
// John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 63-66.
type PhaseAccumulatorEstimator struct {
	mu                               sync.RWMutex
	smoothingLength                  int
	minPeriod                        int
	maxPeriod                        int
	alphaEmaQuadratureInPhase        float64
	alphaEmaPeriod                   float64
	warmUpPeriod                     int
	smoothingLengthPlusHtLengthMin1  int
	smoothingLengthPlus2HtLengthMin2 int
	smoothingLengthPlus2HtLengthMin1 int
	smoothingLengthPlus2HtLength     int
	oneMinAlphaEmaQuadratureInPhase  float64
	oneMinAlphaEmaPeriod             float64
	rawValues                        []float64
	wmaFactors                       []float64
	wmaSmoothed                      []float64
	detrended                        []float64
	deltaPhase                       []float64
	inPhase                          float64
	quadrature                       float64
	count                            int
	smoothedInPhasePrevious          float64
	smoothedQuadraturePrevious       float64
	phasePrevious                    float64
	period                           float64
	isPrimed                         bool
	isWarmedUp                       bool
}

// NewhaseAccumulatorEstimator returns an instnce of the estimator using supplied parameters.
func NewPhaseAccumulatorEstimator(p *CycleEstimatorParams) (*PhaseAccumulatorEstimator, error) {
	err := verifyParameters(p)
	if err != nil {
		return nil, err
	}

	length := p.SmoothingLength
	alphaQuad := p.AlphaEmaQuadratureInPhase
	alphaPeriod := p.AlphaEmaPeriod

	smoothingLengthPlusHtLengthMin1 := length + htLength - 1
	smoothingLengthPlus2HtLengthMin2 := smoothingLengthPlusHtLengthMin1 + htLength - 1
	smoothingLengthPlus2HtLengthMin1 := smoothingLengthPlus2HtLengthMin2 + 1
	smoothingLengthPlus2HtLength := smoothingLengthPlus2HtLengthMin1 + 1

	// These slices will be automatically filled with zeroes.
	wmaSmoothed := make([]float64, htLength)
	detrended := make([]float64, htLength)
	deltaPhase := make([]float64, accumulationLength)
	rawValues := make([]float64, length)
	wmaFactors := make([]float64, length)

	fillWmaFactors(length, wmaFactors)

	return &PhaseAccumulatorEstimator{
		smoothingLength:                  length,
		minPeriod:                        defaultMinPeriod,
		maxPeriod:                        defaultMaxPeriod,
		alphaEmaQuadratureInPhase:        alphaQuad,
		alphaEmaPeriod:                   alphaPeriod,
		warmUpPeriod:                     max(p.WarmUpPeriod, smoothingLengthPlus2HtLength),
		smoothingLengthPlusHtLengthMin1:  smoothingLengthPlusHtLengthMin1,
		smoothingLengthPlus2HtLengthMin2: smoothingLengthPlus2HtLengthMin2,
		smoothingLengthPlus2HtLengthMin1: smoothingLengthPlus2HtLengthMin1,
		smoothingLengthPlus2HtLength:     smoothingLengthPlus2HtLength,
		oneMinAlphaEmaQuadratureInPhase:  1 - alphaQuad,
		oneMinAlphaEmaPeriod:             1 - alphaPeriod,
		rawValues:                        rawValues,
		wmaFactors:                       wmaFactors,
		wmaSmoothed:                      wmaSmoothed,
		detrended:                        detrended,
		deltaPhase:                       deltaPhase,
		period:                           defaultMinPeriod,
	}, nil
}

// SmoothingLength returns the underlying WMA smoothing length in samples.
func (s *PhaseAccumulatorEstimator) SmoothingLength() int {
	return s.smoothingLength
}

// MinPeriod returns the minimal cycle period.
func (s *PhaseAccumulatorEstimator) MinPeriod() int {
	return s.minPeriod
}

// MaxPeriod returns the maximual cycle period.
func (s *PhaseAccumulatorEstimator) MaxPeriod() int {
	return s.maxPeriod
}

// WarmUpPeriod returns the number of updates before the estimator
// is primed (MaxPeriod * 2 = 100).
func (s *PhaseAccumulatorEstimator) WarmUpPeriod() int {
	return s.warmUpPeriod
}

// AlphaEmaQuadratureInPhase returns the value of α (0 < α ≤ 1)
// used in EMA to smooth the in-phase and quadrature components.
func (s *PhaseAccumulatorEstimator) AlphaEmaQuadratureInPhase() float64 {
	return s.alphaEmaQuadratureInPhase
}

// AlphaEmaPeriod returns the value of α (0 < α ≤ 1)
// used in EMA to smooth the instantaneous period.
func (s *PhaseAccumulatorEstimator) AlphaEmaPeriod() float64 {
	return s.alphaEmaPeriod
}

// Count returns the current count value.
func (s *PhaseAccumulatorEstimator) Count() int {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.count
}

// Primed indicates whether an instance is primed.
func (s *PhaseAccumulatorEstimator) Primed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.isWarmedUp
}

// Period returns the current period value.
func (s *PhaseAccumulatorEstimator) Period() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.period
}

// InPhase returns the current InPhase component value.
func (s *PhaseAccumulatorEstimator) InPhase() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.inPhase
}

// Quadrature returns the current Quadrature component value.
func (s *PhaseAccumulatorEstimator) Quadrature() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.quadrature
}

// Detrended returns the current detrended value.
func (s *PhaseAccumulatorEstimator) Detrended() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.detrended[0]
}

// Smoothed returns the current WMA-smoothed value used by underlying Hilbert transformer.
//
// The linear-Weighted Moving Average has a window size of SmoothingLength.
func (s *PhaseAccumulatorEstimator) Smoothed() float64 {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.wmaSmoothed[0]
}

// Update updates the estimator given the next sample value.
func (s *PhaseAccumulatorEstimator) Update(sample float64) { //nolint:funlen
	if math.IsNaN(sample) {
		return
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	push(s.rawValues, sample)

	if s.isPrimed { //nolint:nestif
		if !s.isWarmedUp {
			s.count++
			if s.warmUpPeriod < s.count {
				s.isWarmedUp = true
			}
		}

		// The WMA is used to remove some high-frequency components before detrending the signal.
		push(s.wmaSmoothed, s.wma(s.rawValues))

		amplitudeCorrectionFactor := correctAmplitude(s.period)

		// Since we have an amplitude-corrected Hilbert transformer, and since we want to detrend
		// over its length, we simply use the Hilbert transformer itself as the detrender.
		push(s.detrended, ht(s.wmaSmoothed)*amplitudeCorrectionFactor)

		// Compute both the in-phase and quadrature components of the detrended signal.
		s.quadrature = ht(s.detrended) * amplitudeCorrectionFactor
		s.inPhase = s.detrended[quadratureIndex]

		// Exponential moving average smoothing.
		smoothedInPhase := s.emaQuadratureInPhase(s.inPhase, s.smoothedInPhasePrevious)
		smoothedQuadrature := s.emaQuadratureInPhase(s.quadrature, s.smoothedQuadraturePrevious)
		s.smoothedInPhasePrevious = smoothedInPhase
		s.smoothedQuadraturePrevious = smoothedQuadrature

		// Compute an instantaneous phase.
		phase := instantaneousPhase(smoothedInPhase, smoothedQuadrature, s.phasePrevious)

		// Compute a differential phase.
		push(s.deltaPhase, calculateDifferentialPhase(phase, s.phasePrevious))
		s.phasePrevious = phase

		// Compute an instantaneous period.
		periodPrevious := s.period
		s.period = instantaneousPeriod(s.deltaPhase, periodPrevious)

		// Exponential moving average smoothing of the period.
		s.period = s.emaPeriod(s.period, periodPrevious)
	} else {
		// On (smoothingLength)-th sample we calculate the first
		// WMA smoothed value and begin with the detrender.
		s.count++
		if s.smoothingLength > s.count { // count < 4
			return
		}

		push(s.wmaSmoothed, s.wma(s.rawValues)) // count >= 4

		if s.smoothingLengthPlusHtLengthMin1 > s.count { // count < 10
			return
		}

		amplitudeCorrectionFactor := correctAmplitude(s.period) // count >= 10
		push(s.detrended, ht(s.wmaSmoothed)*amplitudeCorrectionFactor)

		if s.smoothingLengthPlus2HtLengthMin2 > s.count { // count < 16
			return
		}

		s.quadrature = ht(s.detrended) * amplitudeCorrectionFactor // count >= 16
		s.inPhase = s.detrended[quadratureIndex]

		if s.smoothingLengthPlus2HtLengthMin2 == s.count { // count == 16
			s.smoothedInPhasePrevious = s.inPhase
			s.smoothedQuadraturePrevious = s.quadrature

			return
		}

		smoothedInPhase := s.emaQuadratureInPhase(s.inPhase, s.smoothedInPhasePrevious) // count >= 17
		smoothedQuadrature := s.emaQuadratureInPhase(s.quadrature, s.smoothedQuadraturePrevious)
		s.smoothedInPhasePrevious = smoothedInPhase
		s.smoothedQuadraturePrevious = smoothedQuadrature

		phase := instantaneousPhase(smoothedInPhase, smoothedQuadrature, s.phasePrevious)
		push(s.deltaPhase, calculateDifferentialPhase(phase, s.phasePrevious))
		s.phasePrevious = phase

		periodPrevious := s.period
		s.period = instantaneousPeriod(s.deltaPhase, periodPrevious)

		if s.smoothingLengthPlus2HtLengthMin1 < s.count { // count >= 18
			s.period = s.emaPeriod(s.period, periodPrevious)
			s.isPrimed = true
		}
	}
}

func (s *PhaseAccumulatorEstimator) wma(array []float64) float64 {
	value := 0.
	for i := range s.smoothingLength {
		value += s.wmaFactors[i] * array[i]
	}

	return value
}

func (s *PhaseAccumulatorEstimator) emaQuadratureInPhase(value, valuePrevious float64) float64 {
	return s.alphaEmaQuadratureInPhase*value + s.oneMinAlphaEmaQuadratureInPhase*valuePrevious
}

func (s *PhaseAccumulatorEstimator) emaPeriod(value, valuePrevious float64) float64 {
	return s.alphaEmaPeriod*value + s.oneMinAlphaEmaPeriod*valuePrevious
}

func calculateDifferentialPhase(phase, phasePrevious float64) float64 {
	const (
		twoPi         = 2 * math.Pi
		piOver2       = math.Pi / 2
		threePiOver4  = 3 * math.Pi / 4
		minDeltaPhase = twoPi / defaultMaxPeriod
		maxDeltaPhase = twoPi / defaultMinPeriod
	)

	// Compute a differential phase.
	deltaPhase := phasePrevious - phase

	// Resolve phase wraparound from 1st quadrant to 4th quadrant.
	if phasePrevious < piOver2 && phase > threePiOver4 {
		deltaPhase += twoPi
	}

	/*for deltaPhase < 0 {
		deltaPhase += twoPi
	}*/

	// Limit deltaPhase to be within [minDeltaPhase, maxDeltaPhase],
	// i.e. within the bounds of [minPeriod, maxPeriod] sample cycles.
	if deltaPhase < minDeltaPhase {
		deltaPhase = minDeltaPhase
	} else if deltaPhase > maxDeltaPhase {
		deltaPhase = maxDeltaPhase
	}

	return deltaPhase
}

func instantaneousPhase(smoothedInPhase, smoothedQuadrature, phasePrevious float64) float64 {
	// Use arctangent to compute the instantaneous phase in radians.
	phase := math.Atan(math.Abs(smoothedQuadrature / smoothedInPhase))
	if math.IsNaN(phase) || math.IsInf(phase, 0) {
		return phasePrevious
	}

	// Resolve the ambiguity for quadrants 2, 3, and 4.
	if smoothedInPhase < 0 {
		const pi = math.Pi

		if smoothedQuadrature > 0 {
			phase = pi - phase // 2nd quadrant.
		} else if smoothedQuadrature < 0 {
			phase = pi + phase // 3rd quadrant.
		}
	} else if smoothedInPhase > 0 && smoothedQuadrature < 0 {
		const twoPi = 2 * math.Pi

		phase = twoPi - phase // 4th quadrant.
	}

	return phase
}

func instantaneousPeriod(deltaPhase []float64, periodPrevious float64) float64 {
	const twoPi = 2 * math.Pi

	sumPhase := 0.
	period := 0

	for i := range accumulationLength {
		sumPhase += deltaPhase[i]
		if sumPhase >= twoPi {
			period = i + 1

			break
		}
	}

	// Resolve instantaneous period errors.
	if period == 0 {
		return periodPrevious
	}

	return float64(period)
}

File: py/indicators/john_ehlers/hilbert_transformer/phase_accumulator_estimator.py
"""Phase Accumulator cycle estimator implementation."""
import math

from .cycle_estimator import CycleEstimator
from .cycle_estimator_params import CycleEstimatorParams
from .estimator import (
    DEFAULT_MIN_PERIOD, DEFAULT_MAX_PERIOD, HT_LENGTH, QUADRATURE_INDEX,
    ACCUMULATION_LENGTH,
    push, ht, correct_amplitude, fill_wma_factors, verify_parameters,
)


class PhaseAccumulatorEstimator(CycleEstimator):
    """Implements the Hilbert transformer with Phase Accumulation period estimation.

    John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 63-66.
    """

    def __init__(self, p: CycleEstimatorParams) -> None:
        verify_parameters(p)

        length = p.smoothing_length
        alpha_quad = p.alpha_ema_quadrature_in_phase
        alpha_period = p.alpha_ema_period

        sl_ht1 = length + HT_LENGTH - 1          # 10
        sl_2ht2 = sl_ht1 + HT_LENGTH - 1         # 16
        sl_2ht1 = sl_2ht2 + 1                     # 17
        sl_2ht = sl_2ht1 + 1                      # 18

        self._smoothing_length = length
        self._min_period = DEFAULT_MIN_PERIOD
        self._max_period = DEFAULT_MAX_PERIOD
        self._alpha_ema_quadrature_in_phase = alpha_quad
        self._alpha_ema_period = alpha_period
        self._warm_up_period = max(p.warm_up_period, sl_2ht)
        self._sl_ht1 = sl_ht1
        self._sl_2ht2 = sl_2ht2
        self._sl_2ht1 = sl_2ht1
        self._sl_2ht = sl_2ht
        self._one_min_alpha_quad = 1.0 - alpha_quad
        self._one_min_alpha_period = 1.0 - alpha_period

        self._raw_values = [0.0] * length
        self._wma_factors = [0.0] * length
        self._wma_smoothed = [0.0] * HT_LENGTH
        self._detrended = [0.0] * HT_LENGTH
        self._delta_phase = [0.0] * ACCUMULATION_LENGTH
        self._in_phase = 0.0
        self._quadrature = 0.0
        self._count = 0
        self._smoothed_in_phase_prev = 0.0
        self._smoothed_quadrature_prev = 0.0
        self._phase_prev = 0.0
        self._period = float(DEFAULT_MIN_PERIOD)
        self._is_primed = False
        self._is_warmed_up = False

        fill_wma_factors(length, self._wma_factors)

    def smoothing_length(self) -> int:
        return self._smoothing_length

    def min_period(self) -> int:
        return self._min_period

    def max_period(self) -> int:
        return self._max_period

    def warm_up_period(self) -> int:
        return self._warm_up_period

    def alpha_ema_quadrature_in_phase(self) -> float:
        return self._alpha_ema_quadrature_in_phase

    def alpha_ema_period(self) -> float:
        return self._alpha_ema_period

    def count(self) -> int:
        return self._count

    def primed(self) -> bool:
        return self._is_warmed_up

    def period(self) -> float:
        return self._period

    def in_phase(self) -> float:
        return self._in_phase

    def quadrature(self) -> float:
        return self._quadrature

    def detrended(self) -> float:
        return self._detrended[0]

    def smoothed(self) -> float:
        return self._wma_smoothed[0]

    def update(self, sample: float) -> None:
        if math.isnan(sample):
            return

        push(self._raw_values, sample)

        if self._is_primed:
            if not self._is_warmed_up:
                self._count += 1
                if self._warm_up_period < self._count:
                    self._is_warmed_up = True

            push(self._wma_smoothed, self._wma())
            acf = correct_amplitude(self._period)
            push(self._detrended, ht(self._wma_smoothed) * acf)

            self._quadrature = ht(self._detrended) * acf
            self._in_phase = self._detrended[QUADRATURE_INDEX]

            si = self._ema_quad(self._in_phase, self._smoothed_in_phase_prev)
            sq = self._ema_quad(self._quadrature, self._smoothed_quadrature_prev)
            self._smoothed_in_phase_prev = si
            self._smoothed_quadrature_prev = sq

            phase = _instantaneous_phase(si, sq, self._phase_prev)
            push(self._delta_phase, _calculate_differential_phase(phase, self._phase_prev))
            self._phase_prev = phase

            period_prev = self._period
            self._period = _instantaneous_period(self._delta_phase, period_prev)
            self._period = self._ema_period(self._period, period_prev)
        else:
            self._count += 1
            if self._smoothing_length > self._count:
                return

            push(self._wma_smoothed, self._wma())

            if self._sl_ht1 > self._count:
                return

            acf = correct_amplitude(self._period)
            push(self._detrended, ht(self._wma_smoothed) * acf)

            if self._sl_2ht2 > self._count:
                return

            self._quadrature = ht(self._detrended) * acf
            self._in_phase = self._detrended[QUADRATURE_INDEX]

            if self._sl_2ht2 == self._count:
                self._smoothed_in_phase_prev = self._in_phase
                self._smoothed_quadrature_prev = self._quadrature
                return

            si = self._ema_quad(self._in_phase, self._smoothed_in_phase_prev)
            sq = self._ema_quad(self._quadrature, self._smoothed_quadrature_prev)
            self._smoothed_in_phase_prev = si
            self._smoothed_quadrature_prev = sq

            phase = _instantaneous_phase(si, sq, self._phase_prev)
            push(self._delta_phase, _calculate_differential_phase(phase, self._phase_prev))
            self._phase_prev = phase

            period_prev = self._period
            self._period = _instantaneous_period(self._delta_phase, period_prev)

            if self._sl_2ht1 < self._count:  # count >= 18
                self._period = self._ema_period(self._period, period_prev)
                self._is_primed = True

    def _wma(self) -> float:
        value = 0.0
        for i in range(self._smoothing_length):
            value += self._wma_factors[i] * self._raw_values[i]
        return value

    def _ema_quad(self, value: float, value_prev: float) -> float:
        return self._alpha_ema_quadrature_in_phase * value + \
            self._one_min_alpha_quad * value_prev

    def _ema_period(self, value: float, value_prev: float) -> float:
        return self._alpha_ema_period * value + self._one_min_alpha_period * value_prev


def _calculate_differential_phase(phase: float, phase_prev: float) -> float:
    """Computes differential phase with wraparound resolution."""
    TWO_PI = 2.0 * math.pi
    PI_OVER_2 = math.pi / 2.0
    THREE_PI_OVER_4 = 3.0 * math.pi / 4.0
    MIN_DELTA = TWO_PI / DEFAULT_MAX_PERIOD
    MAX_DELTA = TWO_PI / DEFAULT_MIN_PERIOD

    delta = phase_prev - phase

    # Resolve phase wraparound from 1st quadrant to 4th quadrant.
    if phase_prev < PI_OVER_2 and phase > THREE_PI_OVER_4:
        delta += TWO_PI

    if delta < MIN_DELTA:
        delta = MIN_DELTA
    elif delta > MAX_DELTA:
        delta = MAX_DELTA

    return delta


def _instantaneous_phase(
    smoothed_in_phase: float, smoothed_quadrature: float, phase_prev: float
) -> float:
    """Computes instantaneous phase using arctangent with quadrant resolution."""
    phase = math.atan(abs(smoothed_quadrature / smoothed_in_phase)) \
        if smoothed_in_phase != 0.0 else phase_prev

    if math.isnan(phase) or math.isinf(phase):
        return phase_prev

    if smoothed_in_phase < 0:
        if smoothed_quadrature > 0:
            phase = math.pi - phase       # 2nd quadrant
        elif smoothed_quadrature < 0:
            phase = math.pi + phase       # 3rd quadrant
    elif smoothed_in_phase > 0 and smoothed_quadrature < 0:
        phase = 2.0 * math.pi - phase    # 4th quadrant

    return phase


def _instantaneous_period(delta_phase: list[float], period_prev: float) -> float:
    """Computes instantaneous period by phase accumulation."""
    TWO_PI = 2.0 * math.pi
    sum_phase = 0.0
    period = 0

    for i in range(ACCUMULATION_LENGTH):
        sum_phase += delta_phase[i]
        if sum_phase >= TWO_PI:
            period = i + 1
            break

    if period == 0:
        return period_prev

    return float(period)
